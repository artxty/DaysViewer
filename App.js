/* eslint-disable global-require */
import { AppLoading } from 'expo';
import * as Font from 'expo-font';
import PropTypes from 'prop-types';
import { Asset } from 'expo-asset';
import React, { useState } from 'react';
import { Ionicons } from '@expo/vector-icons';
import {
  Platform, StatusBar, StyleSheet, SafeAreaView,
} from 'react-native';
import { Host } from 'react-native-portalize';

import AppNavigator from './src/navigation/AppNavigator';
import appInfoStorage from './src/storage/appInfoStorage';
import { SettingsContextProvider } from './src/storage/settingsContext';

async function loadResourcesAsync() {
  await Promise.all([
    Asset.loadAsync([
      require('./assets/images/robot-dev.png'),
      require('./assets/images/robot-prod.png'),
    ]),
    Font.loadAsync({
      // This is the font that we are using for our tab bar
      ...Ionicons.font,
    }),
  ]);
}

async function detectFirstLaunch() {
  const firstLaunchDate = await appInfoStorage.get('firstLaunch');
  if (firstLaunchDate === null) await appInfoStorage.set('firstLaunch', new Date());
}

function handleLoadingError(error) {
  // In this case, you might want to report the error to your error reporting
  // service, for example Sentry
  console.warn(error);
}

function handleFinishLoading(setLoadingComplete) {
  setLoadingComplete(true);
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
});

export default function App({ skipLoadingScreen }) {
  const [isLoadingComplete, setLoadingComplete] = useState(false);

  if (!isLoadingComplete && !skipLoadingScreen) {
    return (
      <AppLoading
        startAsync={() => {
          loadResourcesAsync();
          detectFirstLaunch();
        }}
        onError={handleLoadingError}
        onFinish={() => handleFinishLoading(setLoadingComplete)}
      />
    );
  }
  return (
    <SettingsContextProvider>
      <SafeAreaView style={styles.container}>
        {Platform.OS === 'ios' && <StatusBar barStyle="dark-content" />}
        <Host>
          <AppNavigator />
        </Host>
      </SafeAreaView>
    </SettingsContextProvider>
  );
}

App.propTypes = {
  skipLoadingScreen: PropTypes.bool,
};

App.defaultProps = {
  skipLoadingScreen: false,
};
