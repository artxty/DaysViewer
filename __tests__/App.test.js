import React from 'react';
import NavigationTestUtils from 'react-navigation/NavigationTestUtils';
import renderer from 'react-test-renderer';

import App from '../App';

jest.mock('expo', () => ({
  AppLoading: 'AppLoading',
}));

jest.mock('../src/navigation/AppNavigator', () => 'AppNavigator');

describe('App', () => {
  beforeEach(() => {
    NavigationTestUtils.resetInternalState();
  });

  it('renders the loading screen', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the root without loading screen', () => {
    const tree = renderer.create(<App skipLoadingScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
