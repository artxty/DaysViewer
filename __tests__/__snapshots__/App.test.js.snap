// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders the loading screen 1`] = `
<AppLoading
  onError={[Function]}
  onFinish={[Function]}
  startAsync={[Function]}
/>
`;

exports[`App renders the root without loading screen 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <AppNavigator />
  </View>
</RCTSafeAreaView>
`;
